File: backend/routers/farmers.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
import logging

from backend.database import get_db
from backend.models import Farmer
from backend.schemas import FarmerCreate, FarmerUpdate, FarmerResponse

router = APIRouter()
logger = logging.getLogger(__name__)


@router.post("/", response_model=FarmerResponse, status_code=status.HTTP_201_CREATED)
def create_farmer(farmer: FarmerCreate, db: Session = Depends(get_db)):
    """Create a new farmer"""
    try:
        # Check if email already exists
        existing_farmer = db.query(Farmer).filter(Farmer.email == farmer.email).first()
        if existing_farmer:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Email already registered"
            )
        
        db_farmer = Farmer(**farmer.model_dump())
        db.add(db_farmer)
        db.commit()
        db.refresh(db_farmer)
        
        logger.info(f"Created farmer: {db_farmer.id}")
        return db_farmer
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating farmer: {str(e)}")
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create farmer"
        )


@router.get("/", response_model=List[FarmerResponse])
def list_farmers(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    """List all farmers with pagination"""
    try:
        farmers = db.query(Farmer).offset(skip).limit(limit).all()
        return farmers
    except Exception as e:
        logger.error(f"Error listing farmers: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve farmers"
        )


@router.get("/{farmer_id}", response_model=FarmerResponse)
def get_farmer(farmer_id: int, db: Session = Depends(get_db)):
    """Get a specific farmer by ID"""
    try:
        farmer = db.query(Farmer).filter(Farmer.id == farmer_id).first()
        if not farmer:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Farmer not found"
            )
        return farmer
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error retrieving farmer {farmer_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve farmer"
        )


@router.put("/{farmer_id}", response_model=FarmerResponse)
def update_farmer(farmer_id: int, farmer_update: FarmerUpdate, db: Session = Depends(get_db)):
    """Update a farmer's information"""
    try:
        db_farmer = db.query(Farmer).filter(Farmer.id == farmer_id).first()
        if not db_farmer:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Farmer not found"
            )
        
        # Update only provided fields
        update_data = farmer_update.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            setattr(db_farmer, field, value)
        
        db.commit()
        db.refresh(db_farmer)
        
        logger.info(f"Updated farmer: {farmer_id}")
        return db_farmer
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating farmer {farmer_id}: {str(e)}")
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to update farmer"
        )


@router.delete("/{farmer_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_farmer(farmer_id: int, db: Session = Depends(get_db)):
    """Delete a farmer"""
    try:
        db_farmer = db.query(Farmer).filter(Farmer.id == farmer_id).first()
        if not db_farmer:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Farmer not found"
            )
        
        db.delete(db_farmer)
        db.commit()
        
        logger.info(f"Deleted farmer: {farmer_id}")
        return None
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting farmer {farmer_id}: {str(e)}")
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to delete farmer"
        )

File: src/routes/farmers.ts
import { Router, Request, Response } from "express";
import { db } from "../data.js";
import { FarmerCreateInput, FarmerUpdateInput } from "../types.js";

export const farmersRouter = Router();

// POST /api/v1/farmers - Create a new farmer
farmersRouter.post("/", (req: Request, res: Response) => {
  const { name, email, phone, address, land_size_acres, crop_types, fpo_id } = req.body as FarmerCreateInput;

  if (!name || !email || !phone) {
    return res.status(422).json({
      detail: [
        { loc: ["body", !name ? "name" : !email ? "email" : "phone"], msg: "Field required", type: "value_error.missing" }
      ]
    });
  }

  // Check if email already exists
  const existing = db.getFarmerByEmail(email);
  if (existing) {
    return res.status(400).json({ detail: "Email already registered" });
  }

  const farmer = db.createFarmer({
    name,
    email,
    phone,
    address: address ?? null,
    land_size_acres: land_size_acres !== undefined ? Number(land_size_acres) : null,
    crop_types: typeof crop_types === "object" ? JSON.stringify(crop_types) : crop_types ?? null,
    fpo_id: fpo_id !== undefined && fpo_id !== null ? Number(fpo_id) : null,
  });

  return res.status(201).json(farmer);
});

// GET /api/v1/farmers - List all farmers with pagination
farmersRouter.get("/", (req: Request, res: Response) => {
  const skip = parseInt(req.query.skip as string, 10) || 0;
  const limit = parseInt(req.query.limit as string, 10) || 100;

  const farmers = db.getFarmers(skip, limit);
  return res.json(farmers);
});

// GET /api/v1/farmers/:farmer_id - Get a specific farmer by ID
farmersRouter.get("/:farmer_id", (req: Request, res: Response) => {
  const farmerId = parseInt(req.params.farmer_id, 10);
  if (isNaN(farmerId)) {
    return res.status(404).json({ detail: "Farmer not found" });
  }

  const farmer = db.getFarmerById(farmerId);
  if (!farmer) {
    return res.status(404).json({ detail: "Farmer not found" });
  }

  return res.json(farmer);
});

// PUT /api/v1/farmers/:farmer_id - Update a farmer's information
farmersRouter.put("/:farmer_id", (req: Request, res: Response) => {
  const farmerId = parseInt(req.params.farmer_id, 10);
  if (isNaN(farmerId)) {
    return res.status(404).json({ detail: "Farmer not found" });
  }

  const existing = db.getFarmerById(farmerId);
  if (!existing) {
    return res.status(404).json({ detail: "Farmer not found" });
  }

  const updates = req.body as FarmerUpdateInput;

  // If email is changing, ensure uniqueness
  if (updates.email && updates.email !== existing.email) {
    const conflict = db.getFarmerByEmail(updates.email);
    if (conflict && conflict.id !== farmerId) {
      return res.status(400).json({ detail: "Email already registered" });
    }
  }

  const updated = db.updateFarmer(farmerId, {
    ...(updates.name !== undefined && { name: updates.name }),
    ...(updates.email !== undefined && { email: updates.email }),
    ...(updates.phone !== undefined && { phone: updates.phone }),
    ...(updates.address !== undefined && { address: updates.address }),
    ...(updates.land_size_acres !== undefined && { land_size_acres: updates.land_size_acres !== null ? Number(updates.land_size_acres) : null }),
    ...(updates.crop_types !== undefined && { crop_types: typeof updates.crop_types === "object" ? JSON.stringify(updates.crop_types) : updates.crop_types }),
    ...(updates.fpo_id !== undefined && { fpo_id: updates.fpo_id !== null ? Number(updates.fpo_id) : null }),
    ...(updates.verified !== undefined && { verified: Boolean(updates.verified) }),
  });

  return res.json(updated);
});

// DELETE /api/v1/farmers/:farmer_id - Delete a farmer
farmersRouter.delete("/:farmer_id", (req: Request, res: Response) => {
  const farmerId = parseInt(req.params.farmer_id, 10);
  if (isNaN(farmerId)) {
    return res.status(404).json({ detail: "Farmer not found" });
  }

  const success = db.deleteFarmer(farmerId);
  if (!success) {
    return res.status(404).json({ detail: "Farmer not found" });
  }

  return res.status(204).send();
});
